### PR TITLE
refactor(git): remove first git checkout as redundant

### PR DESCRIPTION
**Pull request by olegkrivtsov**
## Changes:

Remove first git checkout call as not needed (see https://github.com/renovatebot/renovate/pull/13314#discussion_r781011171)

## Context:

Related to #4294 and PR https://github.com/renovatebot/renovate/pull/13314.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
Needs real terst runs and probably test spec changes

**Comment by JamieMagee** [comment]
It might be the test setup. Here we create `renovate/future_branch` to track the default branch i.e. `main`/`master`/whatever

https://github.com/renovatebot/renovate/blob/af14a38519616e31bd9883e1c27a584127d38827/lib/util/git/index.spec.ts#L42

But the line you're replacing calls `git checkout -B renovate/future_branch origin/renovate/future_branch`

https://github.com/renovatebot/renovate/blob/af14a38519616e31bd9883e1c27a584127d38827/lib/util/git/index.ts#L601

Not 100% sure, but just my thoughts from a glance.

**Comment by olegkrivtsov** [comment]
The problem with the merge test seemed to be related to branch missing locally: https://stackoverflow.com/questions/16862933/how-to-resolve-gits-not-something-we-can-merge-error

I had to modify the test setup to checkout it before merging.

**Review by rarkins — COMMENTED** [review]
This needs testing in a real repo

**Comment by olegkrivtsov** [comment]
> This needs testing in a real repo

@rarkins could you please advice if running Renovate on any real repo would be enough or I need to test some specific merge scenario?

**Comment by rarkins** [comment]
@olegkrivtsov I think you'd need to do a successful branch automerge to verify this. e.g. a repo with at least one passing test so that automergeType=branch + automerge=true can be used

**Comment by olegkrivtsov** [comment]
Hi @rarkins I've tested automerge on this repo: https://github.com/olegkrivtsov/renovate-automerge-test/

First run (it created branch `renovate/httpd-2.x` as expected):

```
$ yarn start
yarn run v1.22.17
$ run-s generate:*
$ node tools/generate-imports.mjs
generating imports
> data/node-js-schedule.json
> node_modules/emojibase-data/en/shortcodes/github.json
$ node -r ts-node/register/transpile-only -- lib/renovate.ts
 INFO: Repository started (repository=olegkrivtsov/renovate-automerge-test)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: Dependency extraction complete (repository=olegkrivtsov/renovate-automerge-test)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "docker-compose": {"fileCount": 1, "depCount": 1},
           "github-actions": {"fileCount": 1, "depCount": 2}
         },
         "total": {"fileCount": 2, "depCount": 3}
       }
 INFO: Branch created (repository=olegkrivtsov/renovate-automerge-test, branch=renovate/httpd-2.x)
       "commitSha": "a528fca"
 INFO: Repository finished (repository=olegkrivtsov/renovate-automerge-test)
       "durationMs": 13456
Done in 38.65s.
```

Second run after tests finished (it automerged branch as expected):

```
$ yarn start
yarn run v1.22.17
$ run-s generate:*
$ node tools/generate-imports.mjs
generating imports
> data/node-js-schedule.json
> node_modules/emojibase-data/en/shortcodes/github.json
$ node -r ts-node/register/transpile-only -- lib/renovate.ts
 INFO: Repository started (repository=olegkrivtsov/renovate-automerge-test)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: Dependency extraction complete (repository=olegkrivtsov/renovate-automerge-test)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "docker-compose": {"fileCount": 1, "depCount": 1},
           "github-actions": {"fileCount": 1, "depCount": 2}
         },
         "total": {"fileCount": 2, "depCount": 3}
       }
 INFO: Branch automerged (repository=olegkrivtsov/renovate-automerge-test, branch=renovate/httpd-2.x)
 INFO: Renovating repository again after automerge result (repository=olegkrivtsov/renovate-automerge-test)
 INFO: Repository started (repository=olegkrivtsov/renovate-automerge-test)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: git fetch completed (repository=olegkrivtsov/renovate-automerge-test)
       "durationMs": 2536
 INFO: Dependency extraction complete (repository=olegkrivtsov/renovate-automerge-test)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "docker-compose": {"fileCount": 1, "depCount": 1},
           "github-actions": {"fileCount": 1, "depCount": 2}
         },
         "total": {"fileCount": 2, "depCount": 3}
       }
 INFO: Repository finished (repository=olegkrivtsov/renovate-automerge-test)
       "durationMs": 5484
Done in 41.04s.
```

Please let me know if this test is enough.

File: lib/util/git/index.spec.ts [hunk 1]
```diff
@@ -223,6 +223,12 @@ describe('util/git/index', () => {
 
   describe('mergeBranch(branchName)', () => {
     it('should perform a branch merge', async () => {
+      const repo = Git(GlobalConfig.get('localDir'));
+      await repo.checkout([
```
Comment (viceice) [inline lib/util/git/index.spec.ts:227]: 🤔 it's currently still unclear to me why this is needed.

**Review by viceice — COMMENTED** [review]
Otherwise LGTM

File: lib/util/git/index.spec.ts (hunk 1 above)
Comment (olegkrivtsov) [reply lib/util/git/index.spec.ts:227]: Hi @viceice I'll try to explain the best way I can. So when I ran the original/unmodified tests, it complained: 

```
FAIL lib/util/git/index.spec.ts (35.928 s, 175 MB heap size)
  ● util/git/index › mergeBranch(branchName) › should perform a branch merge

    merge: renovate/future_branch - not something we can merge

      at Object.action (node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:38:28)
      at PluginStore.exec (node_modules/simple-git/src/lib/plugins/plugin-store.ts:24:29)
      at node_modules/simple-git/src/lib/runners/git-executor-chain.ts:114:40
      at GitExecutorChain.handleTaskData (node_modules/simple-git/src/lib/runners/git-executor-chain.ts:111:14)
      at GitExecutorChain.<anonymous> (node_modules/simple-git/src/lib/runners/git-executor-chain.ts:88:40)
      at fulfilled (node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)
          at runMicrotasks (<anonymous>)
```

The problem with the merge test seemed to be related to branch missing locally (Git can merge two branches if both of them are pulled): https://stackoverflow.com/questions/16862933/how-to-resolve-gits-not-something-we-can-merge-error

I had to modify the test setup to checkout `renovate/future_branch` from `origin/renovate/future_branch` before merging.

Comment (viceice) [reply lib/util/git/index.spec.ts:227]: Ah yes, that would normally done by the commit files function, which isn't called.

**Review by viceice — COMMENTED** [review]
Did you run real test with `persistRepoData` enabled or disabled?

**Comment by olegkrivtsov** [comment]
I guess it is set to default, because I didn't change anything. Here is my `config.js`:

```
module.exports = {
  token: 'XXXXXXXXXXX',
  platform: 'github',
  onboardingConfig: {
    extends: ['config:base'],
  },
  repositories: ['olegkrivtsov/renovate-automerge-test'],
  logFile: '/home/oleg/share/rhys/renovate/error.log',
  logFileLevel: 'debug'
};
```

**Review by viceice — APPROVED** [review]

**Pull request merged**